xdsl/pattern_rewriter.py

```diff
@@ -786,68 +785,128 @@ class PatternRewriteWalker:
+        one use. This is a heuristic based on the fact that single-use operations
+        have more canonicalization opportunities.
+        """
+        for operand in operands:
+            if len(operand.uses) == 1 and isinstance((op := operand.owner), Operation):
+                self._worklist.push(op)
+
+    def _handle_operation_insertion(self, op: Operation) -> None:
+        """Handle insertion of an operation."""
+        if self.apply_recursively:
+            self._worklist.push(op)
+
+    def _handle_operation_removal(self, op: Operation) -> None:
+        """Handle removal of an operation."""
+        if self.apply_recursively:
+            self._add_operands_to_worklist(op.operands)
+        self._worklist.remove(op)
+
+    def _handle_operation_modification(self, op: Operation) -> None:
+        """Handle modification of an operation."""
+        if self.apply_recursively:
+            self._worklist.push(op)
+
+    def _handle_operation_replacement(
+        self, op: Operation, new_results: Sequence[SSAValue | None]
+    ) -> None:
+        """Handle replacement of an operation."""
+        if self.apply_recursively:
+            for result in op.results:
+                for user in result.uses:
+                    self._worklist.push(user.operation)
+
+    def _get_rewriter_listener(self) -> PatternRewriterListener:
+        """
+        Get the listener that will be passed to the rewriter.
+        It will take care of adding operations to the worklist, and calling the
+        listener passed as configuration to the walker.
+        """
+        return PatternRewriterListener(
+            operation_insertion_handler=(
+                self.listener.operation_insertion_handler
+                + [self._handle_operation_insertion]
+            ),
+            operation_removal_handler=(
+                self.listener.operation_removal_handler
+                + [self._handle_operation_removal]
+            ),
+            operation_modification_handler=(
+                self.listener.operation_modification_handler
+                + [self._handle_operation_modification]
+            ),
+            operation_replacement_handler=(
+                self.listener.operation_replacement_handler
+                + [self._handle_operation_replacement]
+            ),
+            block_creation_handler=(self.listener.block_creation_handler),
+        )
+
+    def rewrite_module(self, module: ModuleOp) -> None:
+        """
+        Rewrite operations nested in the given operation by repeatedly applying the
+        pattern.
+        """
+        self.rewrite_op(module)
+
+    def rewrite_op(self, op: Operation) -> None:
+        """
+        Rewrite operations nested in the given operation by repeatedly applying the
+        pattern.
+        """
+        pattern_listener = self._get_rewriter_listener()
 
-    def _rewrite_op(self, op: Operation) -> Operation | None:
+        while True:
```

Comment:
meganit:
```suggestion
        self._populate_worklist(op)
        should_continue = self._process_worklist(pattern_listener)
        if not self.apply_recursively:
            return
        while should_continue:
```
The `while True` here seems a little off to me, and checking for a constant value in a loop

Reply:
It's an arbitrary thing, feel free to ignore

Reply:
Yeah I agree, is this better?
Btw, I was really frustrated to learn that there was no `do/while` in Python :(

Reply:
Yeah do while is useful sometimes